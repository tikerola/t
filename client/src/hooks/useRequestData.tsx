import axios from "axios";
import { useEffect, useState } from "react";

interface HookProps {
  url: string;
  dependencies: [page: number, filter: string];
}

export interface Item {
  id: string;
  color: string[];
  manufacturer: string;
  name: string;
  price: number;
  type: string;
  availability?: string;
}

export interface ResponseData {
  items: Item[];
  numOfItems: number;
}

export const useRequestData = ({ url, dependencies }: HookProps) => {
  const [productData, setProductData] = useState<Item[]>([]);
  const [numOfProducts, setNumOfProducts] = useState(0);

  const [page, filter] = dependencies;

  useEffect(() => {
    const getCategory = async () => {
      try {
        const { data } = await axios["get"]<ResponseData>(url);

        setProductData(data.items);
        setNumOfProducts(data.numOfItems);
      } catch (error) {
        console.log(error);
      }
    };

    getCategory();
  }, [page, filter, url]);

  return { productData, numOfProducts };
};
