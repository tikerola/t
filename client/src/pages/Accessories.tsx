import React from "react";

export const Accessories = (): JSX.Element => {
  return <h1>Hello from Accessories</h1>;
};
