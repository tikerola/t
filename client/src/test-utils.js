import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { lightTheme } from "./styles/theme";
import { Router } from "react-router-dom";

import { createMemoryHistory } from "history";

const history = createMemoryHistory();
const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={lightTheme}>
      <Router history={history}>{children}</Router>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
