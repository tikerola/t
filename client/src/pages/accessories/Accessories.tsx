import React, { useState } from "react";
import { Spinner } from "../../components/spinner/Spinner.styles";
import { useRequestData } from "../../hooks/useRequestData";
import { PRODUCTS_PER_PAGE } from "../common/constants";
import { ProductCategory } from "../common/ProductCategory";

export const Accessories = (): JSX.Element => {
  const [page, setPage] = useState(1);
  const [filter, setFilter] = useState("");

  const { productData, numOfProducts } = useRequestData({
    url: `/products/accessories/${page}?filter=${filter}&ppp=${PRODUCTS_PER_PAGE}`,
    dependencies: [page, filter],
  });

  if (!productData.length && !filter) return <Spinner data-testid="spinner" />;

  return (
    <ProductCategory
      title="Accessories"
      page={page}
      setPage={setPage}
      productData={productData}
      numOfProducts={numOfProducts}
      setFilter={setFilter}
    />
  );
};
