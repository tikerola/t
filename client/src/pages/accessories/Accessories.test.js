import React from "react";
import { render, waitFor } from "../../test-utils";
import { Accessories } from "./Accessories";

import * as hooks from "../../hooks/useRequestData";

const productData = [
  {
    id: "f33561de3a864f951a",
    type: "Accessories",
    name: "EWHHOP ROOM",
    color: ["blue"],
    price: 52,
    manufacturer: "reps",
  },
  {
    id: "0e4772c827c4296592fbd",
    type: "Accessories",
    name: "WEERLEP METROPOLIS RAPTOR",
    color: ["black"],
    price: 98,
    manufacturer: "reps",
  },
  {
    id: "6d39a08b3bcae88a67",
    type: "Accessories",
    name: "DERWEER TYRANNUS BANG",
    color: ["purple"],
    price: 15,
    manufacturer: "abiplos",
  },
];

const data = {
  data: {
    items: productData,
    numOfItems: 15,
  },
};

describe("<Accessories/>", () => {
  it("should match snapshot", async () => {
    const component = render(<Accessories />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should render spinner, when hook return an empty productData array", async () => {
    jest.spyOn(hooks, "useRequestData").mockImplementation(() => ({
      productData: [],
      numOfProducts: 0,
    }));
    const component = render(<Accessories />);
    const spinner = component.getByTestId("spinner");
    expect(spinner).toBeDefined();
  });

  it("should render a product", async () => {
    jest.spyOn(hooks, "useRequestData").mockImplementation(() => ({
      productData: [
        {
          id: "f33561de3a864f951a",
          type: "Accessories",
          name: "EWHHOP ROOM",
          color: ["blue"],
          price: 52,
          manufacturer: "reps",
        },
      ],
      numOfProducts: 15,
    }));
    const component = render(<Accessories />);
    const title = component.getByText(/EWHHOP ROOM/i);
    const manufacturer = component.getByText("reps");

    expect(title).toBeDefined();
    expect(manufacturer).toBeDefined();
  });
});
