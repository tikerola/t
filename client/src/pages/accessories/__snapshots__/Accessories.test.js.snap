// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accessories/> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ EtZUd"
    data-testid="spinner"
  />
</DocumentFragment>
`;
